src/Avalonia.Controls/Presenters/ContentPresenter.cs

```diff
@@ -106,7 +106,8 @@ static ContentPresenter()
         {
             AffectsRender<ContentPresenter>(BackgroundProperty, BorderBrushProperty, BorderThicknessProperty, CornerRadiusProperty);
             AffectsArrange<ContentPresenter>(HorizontalContentAlignmentProperty, VerticalContentAlignmentProperty);
-            AffectsMeasure<ContentPresenter>(BorderThicknessProperty, PaddingProperty);
+            // for ItemsRepeater invalidate measure on a DataContext change; this is for recycled controls to force the Content control to be re-measured
+            AffectsMeasure<ContentPresenter>(BorderThicknessProperty, PaddingProperty, DataContextProperty);
```

Comment:
Hmm, invalidating measure on `DataContext` change here feels a little hackish, though I guess I don't really see a big problem with it in practise.

Reply:
I agree, it is hackish, I would prefer a better alternative.
Perhaps, instead of calling ```InvalidateMeasure``` in ```ViewManager.GetElementFromElementFactory```, call a "new" method to walk down the visual tree of the recycled element and clear the IsMeasureValid flag on all elements.

Reply:
Yeah this is similar to a solution I experimented with. Note that measure only needs to be invalidated when one of the controls in the sub-tree is invalid.

Reply:
I just encountered this problem elsewhere. This is the solution I came up with:

```
private static void InvalidateHack(IControl c)
{
    bool HasInvalidations(IControl c)
    {
        if (!c.IsMeasureValid)
            return true;

        for (var i = 0; i < c.VisualChildren.Count; ++i)
        {
            if (c.VisualChildren[i] is IControl child)
            {
                if (!child.IsMeasureValid || HasInvalidations(child))
                    return true;
            }
        }

        return false;
    }

    void Invalidate(IControl c)
    {
        c.InvalidateMeasure();
        for (var i = 0; i < c.VisualChildren.Count; ++i)
        {
            if (c.VisualChildren[i] is IControl child)
                Invalidate(child);
        }
    }

    if (HasInvalidations(c))
        Invalidate(c);
}
```

Reply:
That looks like it would work